### PR TITLE
fix upnp

### PR DESCRIPTION
**Pull request by ajax16384**
fix upnp xml_parser
fix listen sockets remap

### CODE REVIEW

**Comment by arvidn** [comment]
would you mind adding a test case that would fail without this fix too? (see test/test_xml.cpp)
Do you think this issue exists in RC_1_1 as well?

**Comment by ajax16384** [comment]
ok, a will try to add test/test_xml.cpp with next commit to this pull.

RC_1_1 contains only upnp xmp_parse error (wrong tag strcmp), remap should works well. should i made separate pull request to RC_1_1 ?

**Comment by arvidn** [comment]
> should works well. should i made separate pull request to RC_1_1 ?

that would be great, thanks!

File: src/upnp.cpp [hunk 1]
```diff
@@ -1115,55 +1115,40 @@ void upnp::disable(error_code const& ec)
...
 	}
+}
 
-	struct ip_address_parse_state: public error_code_parse_state
-	{
-		ip_address_parse_state(): in_ip_address(false) {}
-		bool in_ip_address;
-		std::string ip_address;
-	};
+TORRENT_EXTRA_EXPORT void find_ip_address(int type, char const* string, int str_len, ip_address_parse_state& state)
```
Comment (arvidn) [inline src/upnp.cpp]: I believe ``TORRENT_EXTRA_EXPORT`` is redundant here, and in fact adds the risk of a conflict with the header file

File: include/libtorrent/upnp.hpp [hunk 2]
```diff
@@ -120,9 +120,30 @@ struct parse_state
...
 
+struct error_code_parse_state
+{
+	error_code_parse_state(): in_error_code(false), exit(false), error_code(-1) {}
+	bool in_error_code;
+	bool exit;
+	int error_code;
+};
+
+struct ip_address_parse_state: public error_code_parse_state
```
Comment (arvidn) [inline include/libtorrent/upnp.hpp]: the ``public`` here is redundant. for ``struct``, the default inheritance is default.

File: include/libtorrent/upnp.hpp [hunk 3]
```diff
@@ -120,9 +120,30 @@ struct parse_state
...
+{
+	ip_address_parse_state(): in_ip_address(false) {}
```
Comment (arvidn) [inline include/libtorrent/upnp.hpp]: (or ``bool in_ip_address = false;``)

File: src/upnp.cpp [hunk 4]
```diff
@@ -1115,55 +1115,40 @@ void upnp::disable(error_code const& ec)
 	m_socket.close();
 }
 
-namespace
+TORRENT_EXTRA_EXPORT void find_error_code(int type, char const* string, int str_len, error_code_parse_state& state)
```
Comment (arvidn) [inline src/upnp.cpp]: I don't think you should have the ``TORRENT_EXTRA_EXPORT`` here either

**Comment by ajax16384** [comment]
just after this pull request will be squashed to master, i will add another pull request to RC_1_1 with similar changes

**Comment by arvidn** [comment]
lgtm. thanks!

**Pull request merged**